### PR TITLE
added check for cache if bad response from server

### PR DESCRIPTION
**Pull request by margaretgorguissian**
If the remote safety db is down (this happened on 2/15), is it reasonable to check to see if there has been anything cached locally, even if `--cache` is not `true` at the time?

### CODE REVIEW

File: safety/safety.py [hunk 1]
```diff
@@ -15,14 +15,14 @@ class Vulnerability(namedtuple("Vulnerability",
     pass
 
 
-def get_from_cache(db_name):
+def get_from_cache(db_name, site_down=False):
```
Comment (rafaelpivato) [inline safety/safety.py]: The term `site_down` does not align with a cache thing. I mean, we are already getting from cache here. I think we could pass something like `valid_seconds`.

File: safety/safety.py [hunk 2]
```diff
@@ -15,14 +15,14 @@ class Vulnerability(namedtuple("Vulnerability",
...
     if os.path.exists(CACHE_FILE):
         with open(CACHE_FILE) as f:
             try:
                 data = json.loads(f.read())
                 if db_name in data:
                     if "cached_at" in data[db_name]:
-                        if data[db_name]["cached_at"] + CACHE_VALID_SECONDS > time.time():
+                        if site_down or data[db_name]["cached_at"] + CACHE_VALID_SECONDS > time.time():
```
Comment (rafaelpivato) [inline safety/safety.py]: If `valid_seconds` is negative, consider always valid.

File: safety/safety.py [hunk 3]
```diff
@@ -84,6 +84,10 @@ def fetch_database_url(mirror, db_name, key, cached, proxy):
         return data
     elif r.status_code == 403:
         raise InvalidKeyError()
+    else:
+        cached_data = get_from_cache(db_name=db_name, site_down=True)
```
Comment (rafaelpivato) [inline safety/safety.py]: This is a good thing for other HTTP status codes, but we still need to catch **requests** connection exceptions as well.

File: tests/test_safety.py [hunk 4]
```diff
@@ -175,8 +175,8 @@ def test_check_live(self):
             ignore_ids=[],
             proxy={}
         )
-        self.assertEqual(len(vulns), 1)
-
+        self.assertEqual(len(vulns), 1)    
+
```
Comment (rafaelpivato) [inline tests/test_safety.py]: Remove these blanks, please?

File: tests/test_safety.py [hunk 5]
```diff
@@ -204,7 +204,34 @@ def test_check_live_cached(self):
...
+        )
+        self.assertEqual(len(vulns), 1)
+        
+        reqs = StringIO("insecure-package==0.1")
+        packages = util.read_requirements(reqs)
+        # make a second call, where server responds with 400 error, so use
+        # the cache
+        vulns = safety.check(
+            packages=packages,
+            db_mirror="http://pyup.io/",
```
Comment (rafaelpivato) [inline tests/test_safety.py]: Can you use **mock** to avoid letting **requests** performing an actual HTTP request?

**Comment by margaretgorguissian** [comment]
Hi @rafaelpivato, thank you for all your comments! I will address them shortly. Things have been a bit hectic with work and life recently.

**Pull request closed without merging**